fix caller defaultPattern

tbg opened the pull request:
the previous one matches incorrectly when `src` appears before GOPATH.
added TODO for Ben.

bdarnell commented on util/caller/resolver.go:
```diff
@@ -45,8 +45,20 @@ type CallResolver struct {
+	if !ok {
+		panic("unable to look up location")
+	}
+	cur, err := filepath.EvalSymlinks(filepath.Dir(file))
```
Won't this fail if the source tree is not available at runtime?

tschottdorf replied:
yes. I'll remove this (here and below). Hard to imagine someone symlinking around inside of their gopath. And if it happens, it just means that full paths are logged.

bdarnell commented on util/caller/resolver.go:
```diff
@@ -45,8 +45,20 @@ type CallResolver struct {
+	if err != nil {
+		panic(err)
+	}
+	path := filepath.Join(cur, "../../../../..") + sep // $GOPATH/src/
```
All those slashes need to be `sep` instead if we're going to care about `path` versus `filepath`.

tschottdorf replied:
done.

tamird commented on util/caller/resolver.go:
```diff
@@ -45,8 +45,17 @@ type CallResolver struct {
+	}
+	const sep = string(os.PathSeparator)
+	path := filepath.Dir(file)
+	path = filepath.Join(path, strings.Repeat("../", 5)) + sep // $GOPATH/src/
```
This `"../"` looks non-portable to me. I'm not sure yet what the portable equivalent is. Looking.

tamird replied:
Something like:
```
for i := 0; i < 5; i++ {
  path = filepath.Dir(filepath.Clean(path))
}
```

tschottdorf replied:
sorry, I'm just thick. I really wanted `sep`. I like yours more though.

tamird commented on util/caller/resolver.go:
```diff
@@ -45,8 +45,20 @@ type CallResolver struct {
+	for i := 0; i < 5; i++ {
+		path = filepath.Dir(filepath.Clean(path))
+	}
+	return path + sep + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, sep), 3) + "(.*)"
```
I'm not sure, but I think `path` will already have a trailing `sep` here; the final value will not have been `filepath.Clean`'d. You could then inline `sep`.

tamird replied:
actually, you don't need `sep` at all. I think you want `filepath.Join` instead of `strings.Join`

tschottdorf replied:
no, `Dir` always strips the last slash (unless the result is already `/`, which it can't in our case). I was thinking about using `filepath.Join` but it seemed weird since this is a regular expression with characters untypical for a path (plus it looks weird to use it all the time, and inconsistent part of it):

`filepath.Join(path, "[^", "]+", "", .....)`

tamird replied:
Oh, so the `Clean` above isn't necessary either?

Regular expression or not, I'd rather we used `filepath.Join` than our own implementaiton that does exactly the same thing. it looks ok to me:
```
return filepath.Join(path, strings.Repeat(filepath.Join("[^", "]+", ""), 3)) + "(.*)"
```

tamird replied:
ping

tschottdorf replied:
In addition to my discomfort about `filepath.Join` (which may be doing the same thing, but I wouldn't trust that until I had looked at the code), @bdarnell pointed out the need to escape path separators, so the latest version is
```
qSep := regexp.QuoteMeta(sep)
return regexp.MustCompile(regexp.QuoteMeta(path) + qSep + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, qSep), 3) + "(.*)")
```

tamird replied:
seems reasonable

On Tue, Jul 7, 2015 at 8:36 AM, Tobias Schottdorf <notifications@github.com>
wrote:

> In util/caller/resolver.go
> <https://github.com/cockroachdb/cockroach/pull/1632#discussion_r34034148>:
>
> > @@ -45,8 +45,20 @@ type CallResolver struct {
> >  }
> >
> >  // defaultPattern strips src/github.com/organization/project/module/submodule/file.go
> > -// to module/submodule/file.go.
> > -var defaultPattern = "src/" + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, string(os.PathSeparator)), 3) + "(.*)"
> > +// down to module/submodule/file.go.
> > +var defaultPattern = func() string {
> > +	_, file, _, ok := runtime.Caller(0)
> > +	if !ok {
> > +		panic("unable to look up location")
> > +	}
> > +	const sep = string(os.PathSeparator)
> > +	path := filepath.Dir(file)
> > +	// Strip to $GOPATH/src/.
> > +	for i := 0; i < 5; i++ {
> > +		path = filepath.Dir(filepath.Clean(path))
> > +	}
> > +	return path + sep + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, sep), 3) + "(.*)"
>
> In addition to my discomfort about filepath.Join (which may be doing the
> same thing, but I wouldn't trust that until I had looked at the code),
> @bdarnell <https://github.com/bdarnell> pointed out the need to escape
> path separators, so the latest version is
>
> qSep := regexp.QuoteMeta(sep)
> return regexp.MustCompile(regexp.QuoteMeta(path) + qSep + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, qSep), 3) + "(.*)")
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1632/files#r34034148>.
>

bdarnell commented on util/caller/resolver.go:
```diff
@@ -45,8 +45,20 @@ type CallResolver struct {
+	// Strip to $GOPATH/src.
+	for i := 0; i < 5; i++ {
+		path = filepath.Dir(filepath.Clean(path))
+	}
```
Panic if `path` does not end in `/src` at this point (so we can detect if this file is ever moved to a different depth in the package tree).

tschottdorf replied:
done.

bdarnell commented on util/caller/resolver.go:
```diff
@@ -45,8 +45,20 @@ type CallResolver struct {
+	for i := 0; i < 5; i++ {
+		path = filepath.Dir(filepath.Clean(path))
+	}
+	return path + sep + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, sep), 3) + "(.*)"
```
Backslashes are special characters in regular expressions; we need to use `regexp.QuoteMeta` each time we refer to `path` and `sep` in this line. 

We might as well run `regexp.MustCompile` here instead of compiling it every time we create a CallResolver.

tschottdorf replied:
done.

tbg committed:
fix caller defaultPattern

the previous one matches incorrectly when `src` appears before GOPATH.
added TODO for Ben.

bdarnell commented:
LGTM
